### PR TITLE
CNV-19689: Add note about resource quotas

### PR DESCRIPTION
**Pull request by bgaydosrh**
For 4.8, 4.9, 4.10, 4.11, and 4.12.

Jira: https://issues.redhat.com/browse/CNV-19689
BZ: https://bugzilla.redhat.com/show_bug.cgi?id=2104430

Direct doc preview link: https://48252--docspreview.netlify.app/openshift-enterprise/latest/virt/virtual_machines/advanced_vm_management/virt-working-with-resource-quotas-for-vms.html

Tagging @fabiand  and @stu-gott  for Code Review
Tagging Oded Ramraz to assign a QE in Jira.

Thakns
Bob

### CODE REVIEW

File: modules/virt-about-resource-quota-limits.adoc [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+    spec:
+      domain:
+        …
+        resources:
+          requests:
+            memory: 128Mi
+          limits:
+            memory: 256Mi  <1>
+----
+<1> `resources/limits` is twice the size of `resource/requests`
```
Comment (stu-gott) [inline modules/virt-about-resource-quota-limits.adoc:31]: I think this caveat can be misunderstood when written this way. Memory limits in general need to be larger than requests. currently that delta is about 100M and has slowly changed over time. In this example the limit is presented as twice the request, but to state that without context might imply there's a generalized rule to always double--which is not the case for larger requests.

**Review by stu-gott — COMMENTED** [review]
One small suggestion, otherwise this looks good. thanks!

File: modules/virt-about-resource-quota-limits.adoc (hunk 1 above)
Comment (bgaydosrh) [reply modules/virt-about-resource-quota-limits.adoc]: Hi @stu-gott - I had the same thought, lol. I have updated accordingly.

**Review by stu-gott — COMMENTED** [review]
LGTM for content

**Comment by stu-gott** [comment]
I've never seen that before. I reviewed so fast that github got the sequence out of order. My LGTM applies to this point in time.

File: modules/virt-about-resource-quota-limits.adoc [hunk 2]
```diff
@@ -0,0 +1,31 @@
+// Module included in the following assemblies:
+//
+// * virt/virtual_machines/advanced_vm_management/virt-specifying-nodes-for-vms.adoc
+
+:_content-type: CONCEPT
+[id="virt-about-resource-quota-limits_{context}"]
+= About resource quota limits
+
+You can only set `ResourceQuotas` for virtual machines if the resource quota has a set of requests. If the resource quota contains both requests and limits, the request limits on the virtual machine must be set.
```
Comment (dshchedr) [inline modules/virt-about-resource-quota-limits.adoc:9]: > If the resource quota contains both requests and limits, the request limits on the virtual machine must be set.
This sentence sounds confusing for me.. 

I think there should be AND:
` the requests AND limits on the virtual machine must be set.`

Comment (dshchedr) [reply modules/virt-about-resource-quota-limits.adoc:9]: or:
`the resource limits on the virtual machine must be set.`

Comment (stu-gott) [reply modules/virt-about-resource-quota-limits.adoc:9]: Well spotted. I think that was actually supposed to say "resource limits".

Comment (bgaydosrh) [reply modules/virt-about-resource-quota-limits.adoc]: Thanks @dshchedr . Updated.

Comment (stu-gott) [reply modules/virt-about-resource-quota-limits.adoc]: Well spotted @dshchedr, I believe this should have said "resource limits"

**Comment by dshchedr** [comment]
Thanks, 
/LGTM

File: modules/virt-about-resource-quota-limits.adoc (hunk 2 above)
Comment (bgaydosrh) [reply modules/virt-about-resource-quota-limits.adoc]: +1, thanks to you both.

File: modules/virt-about-resource-quota-limits.adoc [hunk 3]
```diff
@@ -0,0 +1,31 @@
+// Module included in the following assemblies:
+//
+// * virt/virtual_machines/advanced_vm_management/virt-specifying-nodes-for-vms.adoc
+
+:_content-type: CONCEPT
+[id="virt-about-resource-quota-limits_{context}"]
+= About resource limits
+
+You can only set `ResourceQuotas` for virtual machines if the resource quota has a set of requests. If the resource quota contains both requests and limits, the resource limits on the virtual machine must be set.
```
Comment (ousleyp) [inline modules/virt-about-resource-quota-limits.adoc:9]: - I am confused by this intro, as well as its location in this assembly (which doesn't mention resource quotas).

- I think this should be a procedure rather than a concept. 

- After reviewing the bug, I would reword the intro to something like this:
> Resource quotas that only use requests automatically work with virtual machines (VMs). If your resource quota uses limits, you must manually set resource limits on VMs. Resource limits must be at least 100 MiB larger than resource requests.
- Re: s/megabytes/MiB: the `Mi` in the YAML makes me think that it is supposed to be MiB (mebibytes). The ISG also seems to recommend just writing `MiB` instead of spelling it out: https://www.ibm.com/docs/en/ibm-style?topic=measurement-units

- I think there should be a procedure step or two; something like this:

```
. Set limits for a VM by editing the `VirtualMachine` manifest. For example:
+
<yaml example>

. <step to apply changes>
```

**Review by ousleyp — COMMENTED** [review]
I think this would benefit from a bit of rework. Please let me know if you have any questions about my comments. Thanks!

**Comment by bergerhoffer** [comment]
The enterprise-4.12 label has been added to this PR.

This is because your PR targets the `main` branch and is labeled for enterprise-4.11. And any PR going into `main` must also target the latest version branch (enterprise-4.12).

If the update in your PR does NOT apply to version 4.12 onward, please re-target this PR to go directly into the appropriate version branch or branches (enterprise-4.x) instead of main.

**Comment by bgaydosrh** [comment]
> The enterprise-4.12 label has been added to this PR.
> 
> This is because your PR targets the `main` branch and is labeled for enterprise-4.11. And any PR going into `main` must also target the latest version branch (enterprise-4.12).
> 
> If the update in your PR does NOT apply to version 4.12 onward, please re-target this PR to go directly into the appropriate version branch or branches (enterprise-4.x) instead of main.

Thanks, 4.12 label applied

File: virt/virtual_machines/advanced_vm_management/virt-specifying-nodes-for-vms.adoc [hunk 4]
```diff
@@ -8,6 +8,8 @@ toc::[]
 
 You can place virtual machines (VMs) on specific nodes by using node placement rules.
 
+include::modules/virt-about-resource-quota-limits.adoc[leveloffset=+1]
```
[the remark this answers is not in the document]
Comment (fabiand) [reply virt/virtual_machines/advanced_vm_management/virt-specifying-nodes-for-vms.adoc:11]: After revisiting the 4.10 docs, I'd suggest that inside the "Advanced Virtual Machine management" section, we add a new section  called "Working with ResourceQuotas".

File: modules/virt-about-resource-quota-limits.adoc [hunk 5]
```diff
@@ -0,0 +1,31 @@
+// Module included in the following assemblies:
+//
+// * virt/virtual_machines/advanced_vm_management/virt-specifying-nodes-for-vms.adoc
+
+:_content-type: CONCEPT
+[id="virt-about-resource-quota-limits_{context}"]
+= About resource limits
```
[the remark this answers is not in the document]
Comment (fabiand) [reply modules/virt-about-resource-quota-limits.adoc:7]: However it is worded, but the title should refer to working with ResourceQuotas

File: modules/virt-about-resource-quota-limits.adoc (hunk 3 above)
Comment (fabiand) [reply modules/virt-about-resource-quota-limits.adoc:9]: Pan, +1 on your suggested wording.

File: modules/virt-about-resource-quota-limits.adoc [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+apiVersion: kubevirt.io/v1
+kind: VirtualMachine
+metadata:
+  name: with-limits
+spec:
+  running: false
+  template:
+    spec:
+      domain:
+        …
```
[the remark this answers is not in the document]
Comment (fabiand) [reply modules/virt-about-resource-quota-limits.adoc:24]: Whatever is needed :)

File: modules/virt-about-resource-quota-limits.adoc [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+        resources:
+          requests:
+            memory: 128Mi
+          limits:
+            memory: 256Mi  <1>
+----
+<1> `resources/limits` is twice the size of `resource/requests`, meeting the `100Mi` requirement.
```
[the remark this answers is not in the document]
Comment (fabiand) [reply modules/virt-about-resource-quota-limits.adoc:31]: +1 on Pans wording

**Review by fabiand — COMMENTED** [review]
Sorry for taking some time to get to this, but I also had to wrap my head around it again as well.

**Comment by fabiand** [comment]
@bgaydosrh anything else that I can help with?

@ousleyp

**Comment by bgaydosrh** [comment]
> @bgaydosrh anything else that I can help with?
> 
> @ousleyp

No, thanks, @fabiand.

File: virt/virtual_machines/advanced_vm_management/virt-specifying-nodes-for-vms.adoc (hunk 4 above)
[the remark this answers is not in the document]
Comment (bgaydosrh) [reply virt/virtual_machines/advanced_vm_management/virt-specifying-nodes-for-vms.adoc]: Thanks @ousleyp and @fabiand and apologies for taking so long to wrap this up :-)

After also talking with @stu-gott to better understand resource quotas, it does seem like this topic in OpenShift covers many of the basics:  https://docs.openshift.com/container-platform/4.11/applications/quotas/quotas-setting-per-project.html. The difference, as I understand it, is in CNV, we only work with quotas for virtual machines as opposed to pods.

If this is true, then your suggestion to make this a separate topic makes a lot of sense. So, my idea is to create another topic as you both have suggested and include the module pertaining to limits with respect to CNV.

@fabiand, if you agree to this, I can work with @ousleyp on the how detailed to make this topic. It appears there might be several modules in the linked OCP assembly that I can reuse in regards to creating the YAML, etc. but we obviously need to include an example that is CNV/VM-specific.

Pan, I also am a little unsure about my opening wording for this new assembly, so perhaps you can suggest something. I agree with all your other comments.... the concept and placement of the topic and possible overlap with the OCP doc is what has been a little harder to grasp, for me.

For this pass I am adding the new assembly and a pointer in the topic map. When I get your replies, I can proceed with the rest of the changes. I just want to verify I am heading in the right direction first. :-)

And I thank you for your patience!

Bob

File: modules/virt-about-resource-quota-limits.adoc (hunk 6 above)
[the remark this answers is not in the document]
Comment (bgaydosrh) [reply modules/virt-about-resource-quota-limits.adoc:24]: Done, thanks I was not aware of the `#` but that makes sense.

File: modules/virt-about-resource-quota-limits.adoc (hunk 7 above)
[the remark this answers is not in the document]
Comment (bgaydosrh) [reply modules/virt-about-resource-quota-limits.adoc:31]: Thanks!

File: modules/virt-about-resource-quota-limits.adoc (hunk 5 above)
[the remark this answers is not in the document]
Comment (bgaydosrh) [reply modules/virt-about-resource-quota-limits.adoc]: I changed the ID and title to "Setting resource quota limits" and made this a PROCEDURE module called by the "Working with resource quotas" assembly.

@ousleyp - Do you think I should retain "in {VirtProductName}" in the assembly and module titles to differentiate from OCP or do you think that's just redundant?

File: modules/virt-about-resource-quota-limits.adoc (hunk 3 above)
Comment (bgaydosrh) [reply modules/virt-about-resource-quota-limits.adoc:9]: Thanks, Pan. Much clearer.

File: virt/virtual_machines/advanced_vm_management/virt-specifying-nodes-for-vms.adoc (hunk 4 above)
[the remark this answers is not in the document]
Comment (ousleyp) [reply virt/virtual_machines/advanced_vm_management/virt-specifying-nodes-for-vms.adoc]: I'd definitely suggest keeping the procedure in its own module; assemblies without any modules are rare edge cases and usually created out of necessity/as a workaround for something like the xref-in-modules ban

File: modules/virt-about-resource-quota-limits.adoc (hunk 5 above)
[the remark this answers is not in the document]
Comment (ousleyp) [reply modules/virt-about-resource-quota-limits.adoc]: I don't think we need "in {VirtProductName}" but I wouldn't be opposed to you adding something like "...for virtual machines" to a title. I'm kind of ambivalent, though, since it's in the CNV docs

File: modules/virt-about-resource-quota-limits.adoc [hunk 8]
```diff
@@ -0,0 +1,35 @@
+// Module included in the following assemblies:
+//
+// * virt/virtual_machines/advanced_vm_management/virt-specifying-nodes-for-vms.adoc
+
+:_content-type: PROCEDURE
+[id="virt-setting-resource-quota-limits_{context}"]
```
Comment (ousleyp) [inline modules/virt-about-resource-quota-limits.adoc:6]: Please rename the module file itself to match the new title/ID (and of course, update it in the assembly when you do). Thanks!

File: modules/virt-about-resource-quota-limits.adoc [hunk 9]
```diff
@@ -0,0 +1,35 @@
+// Module included in the following assemblies:
+//
+// * virt/virtual_machines/advanced_vm_management/virt-specifying-nodes-for-vms.adoc
```
Comment (ousleyp) [inline modules/virt-about-resource-quota-limits.adoc:3]: Please update this to the new assembly path that you created

File: virt/virtual_machines/advanced_vm_management/virt-working-with-resource-quotas.adoc [hunk 10]
```diff
@@ -0,0 +1,11 @@
+:_content-type: ASSEMBLY
+[id="virt-working-with-resource-quotas"]
+= Working with resource quotas in {VirtProductName}
+include::_attributes/common-attributes.adoc[]
+:context: virt-working-with-resource-quotas
+
+toc::[]
+
+Resource quotas are implemented in {VirtProductName} similarly to how they are implemented in {product-title}. However, in (VirtProductName}, you are creating and managing resource quotas only for virtual machines.
```
Comment (ousleyp) [inline virt/virtual_machines/advanced_vm_management/virt-working-with-resource-quotas.adoc:9]: I might reword this intro to something simpler, since the 'meat' of this doc change is in the procedure module. You can have an assembly intro that is as simple as "Create and manage resource quotas for virtual machines."

Suggestion: we could also include an xref to the OCP resource quota docs somewhere, if that would be useful.

File: virt/virtual_machines/advanced_vm_management/virt-specifying-nodes-for-vms.adoc (hunk 4 above)
[the remark this answers is not in the document]
Comment (bgaydosrh) [reply virt/virtual_machines/advanced_vm_management/virt-specifying-nodes-for-vms.adoc]: Cool.

File: modules/virt-about-resource-quota-limits.adoc (hunk 5 above)
[the remark this answers is not in the document]
Comment (bgaydosrh) [reply modules/virt-about-resource-quota-limits.adoc]: I'll probably include "for VMs" just for findability.

File: modules/virt-about-resource-quota-limits.adoc (hunk 8 above)
Comment (bgaydosrh) [reply modules/virt-about-resource-quota-limits.adoc:6]: Done.

File: modules/virt-about-resource-quota-limits.adoc (hunk 9 above)
Comment (bgaydosrh) [reply modules/virt-about-resource-quota-limits.adoc:3]: Doh! Thanks.

File: virt/virtual_machines/advanced_vm_management/virt-working-with-resource-quotas.adoc (hunk 10 above)
Comment (bgaydosrh) [reply virt/virtual_machines/advanced_vm_management/virt-working-with-resource-quotas.adoc]: Went with the simple recommendation. I thought about mentioning limits for a more comprehensive abstract sentence but I also don't want to limit the scope of this assembly.

File: virt/virtual_machines/advanced_vm_management/virt-working-with-resource-quotas-for-vms.adoc [hunk 11]
```diff
@@ -0,0 +1,17 @@
...
+include::_attributes/common-attributes.adoc[]
+:context: virt-working-with-resource-quotas-for-vms
+
+toc::[]
+
+Create and manage resource quotas for virtual machines.
+
+include::modules/virt-setting-resource-quota-limits-for-vms.adoc[leveloffset=+1]
+
+.Additional resources
```
Comment (ousleyp) [inline virt/virtual_machines/advanced_vm_management/virt-working-with-resource-quotas-for-vms.adoc:13]: Please use the add'l resources formatting for assemblies:

```
[role="_additional-resources"]
[id="additional-resources_virt-working-with-resource-quotas-for-vms"]
== Additional resources
```

**Review by ousleyp — COMMENTED** [review]
Just one nit re: additional resources formatting; otherwise this LGTM

File: virt/virtual_machines/advanced_vm_management/virt-working-with-resource-quotas-for-vms.adoc (hunk 11 above)
Comment (bgaydosrh) [reply virt/virtual_machines/advanced_vm_management/virt-working-with-resource-quotas-for-vms.adoc:15]: Done. Thanks.

Comment (ousleyp) [reply virt/virtual_machines/advanced_vm_management/virt-working-with-resource-quotas-for-vms.adoc:15]: Thanks for adding the `role` and `id` lines! One last thing, though: you still have `.Additional resources` in line 15. That should be changed to `== Additional resources`.

**Review by ousleyp — COMMENTED** [review]
I left one comment regarding the additional resources section. Please let me know when you've fixed that last thing and I'll go ahead with merging.

File: virt/virtual_machines/advanced_vm_management/virt-working-with-resource-quotas-for-vms.adoc (hunk 11 above)
Comment (bgaydosrh) [reply virt/virtual_machines/advanced_vm_management/virt-working-with-resource-quotas-for-vms.adoc:15]: Sorry about that. Should be good now :-)
